test/browser-reporters/html.spec.js

```diff
@@ -0,0 +1,1918 @@
+        runner.emit('suite', rootSuite.suites[2]);
+        runner.emit('test', rootSuite.suites[2].tests[0]);
+        rootSuite.suites[2].tests[0].duration = 0;
+        rootSuite.suites[2].tests[0].state = 'failed';
+        try {
+          failingTest();
+          throw new Error('Failing test did not throw assertion');
+        } catch (assertion) {
+          runner.emit('fail', rootSuite.suites[2].tests[0], assertion);
+        }
+        runner.emit('test end', rootSuite.suites[2].tests[0]);
+        runner.emit('suite end', rootSuite.suites[2]);
+        runner.emit('suite end', rootSuite);
+        runner.emit('end');
+        click(output.getElementsByClassName('failures').item(0).getElementsByTagName('a').item(0));
+        var suites = output.getElementsByClassName('suite');
+        for (var index = 0; index < suites.length; index += 1) {
+          if (!suites[index].getElementsByClassName('test fail').length) {
+            assertDisplayed(suites[index], false);
+          }
+        }
+        restoreOutput();
+      } catch (error) {
+        try {
+          restoreOutput();
+        } catch (ignore) {}
+        throw error;
+      }
+    });
+
+    it('should not hide suites with failed tests', function () {
+      try {
+        var output = stubOutput();
+        var runner = new Runner(4);
+        new HTML(runner); // eslint-disable-line no-new
+        var rootSuite = new Suite('', new Context());
+        Suite.create(rootSuite, 'passing');
+        rootSuite.suites[0].addTest(new Test('passing', function () { /* content here */; })); // eslint-disable-line no-extra-semi
+        Suite.create(rootSuite, 'pending');
+        rootSuite.suites[1].addTest(new Test('pending'));
+        rootSuite.suites[1].addTest(new Test('skipped at runtime', function () { this.skip(); }));
+        Suite.create(rootSuite, 'failing');
+        var failingTest = function () { throw new Error('fail test'); };
+        rootSuite.suites[2].addTest(new Test('failing', failingTest));
+        runner.emit('start');
+        runner.emit('suite', rootSuite);
+        runner.emit('suite', rootSuite.suites[0]);
+        runner.emit('test', rootSuite.suites[0].tests[0]);
+        rootSuite.suites[0].tests[0].duration = 0;
+        rootSuite.suites[0].tests[0].state = 'passed';
+        runner.emit('pass', rootSuite.suites[0].tests[0]);
+        runner.emit('test end', rootSuite.suites[0].tests[0]);
+        runner.emit('suite end', rootSuite.suites[0]);
+        runner.emit('suite', rootSuite.suites[1]);
+        // no 'test' event is emitted for pending tests
+        runner.emit('pending', rootSuite.suites[1].tests[0]);
+        runner.emit('test end', rootSuite.suites[1].tests[0]);
+        runner.emit('test', rootSuite.suites[1].tests[1]);
+        rootSuite.suites[1].tests[1].pending = true;
+        rootSuite.suites[1].tests[1].duration = 0;
+        runner.emit('pending', rootSuite.suites[1].tests[1]);
```

Comment:
Could have pulled in the real `Runner`, but then this would have been more like an integration test than a unit test... (I made a tradeoff to reuse the real `Suite` and `Test` objects mainly because I felt that the code would be too bloated if I tried to recreate them completely but too brittle if I tried to recreate just the bits and pieces that the HTML reporter currently happens to be relying upon.)